crypto/bio/bss_dgram.c

```diff
@@ -808,6 +992,627 @@ static int dgram_puts(BIO *bp, const char *str)
+                BIO_MSG_N(msg, i + j).flags    = 0;
+            }
+            num_done += ret;
+        } else if (ret < 0) {
+            return num_done > 0 ? num_done : PACK_ERRNO(get_last_socket_error());
+        }
+
+        /*
+         * If we didn't send as many messages as we asked for, OS buffers must
+         * be full, so stop here.
+         */
+        if ((size_t)ret < jlim)
+            break;
+    }
+
+    return num_done;
+
+#elif M_METHOD == M_METHOD_RECVMSG
+    /*
+     * If sendmsg is available, we emulate sendmmsg using multiple calls.
+     */
+    for (i = 0; i < num_msg; ++i) {
+        translate_msg(&mh, &iov, control, &BIO_MSG_N(msg, i));
+
+        if (BIO_MSG_N(msg, i).local != NULL) {
+            if (!have_local_enabled)
+                return i > 0 ? (ossl_ssize_t)i : -1;
+
+            if (pack_local(b, &mh, BIO_MSG_N(msg, i).local) < 1)
+                return i > 0 ? (ossl_ssize_t)i : -1;
+        }
+
+        ret = sendmsg(b->num, &mh, sysflags);
+        if (ret <= 0)
+            break;
+
+        BIO_MSG_N(msg, i).data_len = ret;
+        BIO_MSG_N(msg, i).flags    = 0;
+    }
+
+    return i > 0 ? (ossl_ssize_t)i : PACK_ERRNO(get_last_socket_error());
+
+#elif M_METHOD == M_METHOD_WSARECVMSG || M_METHOD == M_METHOD_RECVFROM
+# if M_METHOD == M_METHOD_WSARECVMSG
+    if (bio_WSASendMsg != NULL) {
+        /* WSASendMsg-based implementation for Windows. */
+        translate_msg_win(&wmsg, &wbuf, control, msg);
+
+        if (msg[0].local != NULL) {
+            if (!have_local_enabled)
+                return -1;
+
+            if (pack_local(b, &wmsg, msg[0].local) < 1)
+                return -1;
+        }
+
+        ret = WSASendMsg((SOCKET)b->num, &wmsg, 0, &num_bytes_sent, NULL, NULL);
+        if (ret < 0)
+            return PACK_ERRNO(get_last_socket_error());
+
+        msg[0].data_len = num_bytes_sent;
+        msg[0].flags    = 0;
+        return 1;
+    }
+# endif
+
+    /*
+     * Fallback to sendto and send a single message.
+     */
+    if (msg[0].local != NULL)
+        /*
+         * We cannot set the local address if using sendto
+         * so fail in this case
+         */
+        return -1;
+
+    ret = sendto(b->num, msg[0].data,
+# if defined(OPENSSL_SYS_WINDOWS)
+                 (int)msg[0].data_len,
+# else
+                 msg[0].data_len,
+# endif
+                 sysflags,
+                 msg[0].peer != NULL ? &msg[0].peer->sa : NULL,
+                 msg[0].peer != NULL ? sizeof(*msg[0].peer) : 0);
+    if (ret <= 0)
+        return PACK_ERRNO(get_last_socket_error());
+
+    msg[0].data_len = ret;
+    msg[0].flags    = 0;
+    return 1;
+
+#else
+    ERR_raise(ERR_LIB_BIO, BIO_R_UNSUPPORTED_METHOD);
+    return -2;
+#endif
+}
+
+static ossl_ssize_t dgram_recvmmsg(BIO *b, BIO_MSG *msg, size_t stride, size_t num_msg, uint64_t flags)
+{
+#if M_METHOD != M_METHOD_NONE
+    int ret;
+#endif
+#if M_METHOD == M_METHOD_RECVMMSG
+# define BIO_MAX_MSGS_PER_CALL   64
```

Comment:
Why doesn't this end up being defined twice? Didn't we already define this in dgram_sendmmsg above?